# Pandas

The first part of any data project is the handling and manipulation of data 
in order to prepare for data analysis. This is where the library Pandas comes in to assist the process.

Pandas name comes from the word "Panel Data" - Which is a multidimension date of measurements of time. Pandas is an open source library to assist with data manipulation.

Features of Pandas:

 * Series Objects and Data Frames
     * Allows to deal with 1 and 2 dimensional data
 * Handling of Missing Data
     * Presented as NaN
representation:
 * Data Alignment
     * Data will be aligned to a set of labels or Pandas objects will allocate the data to a set of labels itself.
 * Functions to Group the data:
     * Split, Combine  and apply opperations can be used here
 * Slicing, Indexing and Subsetting:
     * Can be done to the data set
 * Merging and Joining:
    * Allows us to have an intellgent method to combine multiple datasets. 
 * Reshape:
   * Allows us to modify how the dataset looks for analysis.
 * Hierachical Labelling of Axes
 * Robust input output tool
     * Can input from multiple sources
 * Time series specific functionality
 

## Pandas Vs Numpy

In Python there are two different Libraries that handle any data requests. We will be focusing on Pandas but it may be useful to see the difference between Numpy and Pandas:
 
 * Pandas works better with 500k+ rows than Numpy
 * Pandas Series is more flexible than Numpy Arrays

## How to Import Pandas into Python

A Simple command is used to import pandas, much like the other librarys that may be Required:

In [0]:
import pandas as pd

In the command above we are importing the pandas library and we are allocating it a name that we will be able to use in the rest of the code (pd).

## Datasets in Pandas

Pandas handles anything between a one dimensional to a multiple dimension dataset.

If the data is one dimensional, Pandas will treat it as a Series object, If it is two dimension it will be treated as a DataFrame and finally multi-dimensional is treated as Panel data.

### One Dimensional - Series data

A series object is a Single column Array that can hold any kind of data Type. It can be mixed or similar data.

To Create a series object we will use the data of (1,2,3,4).

Let's see how to create this into a series and then print the data:

In [40]:
data = [1,2,3,4]
Series1 = pd.Series(data)
Series1

0    1
1    2
2    3
3    4
dtype: int64

Please note that the data now has a 0 - 3 running down the side depicting the row number (Python is a language that is 0 indexed). Also remember that Python is case sensitive so watch all cases in the calling of data and functions. 

To check the type of the object we can use the type() function:

In [41]:
type(Series1)

pandas.core.series.Series

It is nice that we have an index of 0 - 3 but we may not always want this and rather prefer some different details. This can be done with including the index option in the series function. Take note of the index after we run the function again. Let's see how below:

In [42]:
Series1 = pd.Series(data,index=['a','b','c','d'])
Series1

a    1
b    2
c    3
d    4
dtype: int64

The Series data can also be created using array's, Dictionary or Scalar or constant value to create a Series with it.

### Two Dimensional - DataFrame

A DataFrame is a two dimensional labeled data structure with columns with potentially different types. They have the following features:

* Different Columns Types
* Mutable size
* Labeled axes
* Arithmic operations on rows and columns

But the first important step is the creation of the DataFrame, We will now create a new dataset and turn it into a dataframe below:

In [43]:
data = [1,2,3,4,5]
df = pd.DataFrame(data)
df

0
0  1
1  2
2  3
3  4
4  5



```
# This is formatted as code
```

We can create the dataframe from a dictionary as well, for this example we will create a dictionary of the number of fruits we have and then convert it into a dataframe.

In [44]:
dictionary = {'fruits':['apples','banana','mangoes'],'count':[10,20,15]}
df = pd.DataFrame(dictionary)
df

fruits  count
0   apples     10
1   banana     20
2  mangoes     15

We can now see that the columns have the labels included that were set in the dictionary as well as the data that is both the numerical and string values.

The next thing to look at is how to create a DataFrame from a series dataset:

In [45]:
Series = pd.Series([6,12],index=['a','b'])
df = pd.DataFrame(Series)
df

0
a   6
b  12

Now we can see that the one dimensional Series was converted into a two dimensional dataframe with an index and column. If you think about the original Series example you will see that the new dataset now has a column name.

Most data formats can be converted into a pandas dataframe in a similar fashion to all of the examples.

## Merge, Join and concatenate functions

When we are working with multiple datasets we may get to a point were we will need to combine the datasets. There are three different functions that allow us to do this.

We will go through all of the three functions to see the differences. The first step in this example is to create two dataframes and start working with the Merge Function:

In [46]:
#Data frame 1 creation

player = ['Player1','Player2','Player3']
point=[8,9,6]
title=['Game1','Game2','Game3']

df1 = pd.DataFrame({'Player':player,'Points':point,'Title':title})
df1

Player  Points  Title
0  Player1       8  Game1
1  Player2       9  Game2
2  Player3       6  Game3

In [47]:
#Dataframe 2 creation

player = ['Player1','Player5','Player6']
power = ['Punch','Kick','Elbow']
title=['Game1','Game5','Game6']
df2 = pd.DataFrame({'Player':player,'Power':power,'Title':title})
df2

Player  Power  Title
0  Player1  Punch  Game1
1  Player5   Kick  Game5
2  Player6  Elbow  Game6

The Merge function works with the data inside both of the DataFrames to join the data, this is especially useful if there are some unique identifyers that can be used to connect the data.

One of the first types of Merge is the Inner Merge. This Merge looks for the common values between the two DataFrames and only outputs the values that are common between the two dataframes.

In [48]:
df1.merge(df2, on='Title',how='inner')

Player_x  Points  Title Player_y  Power
0  Player1       8  Game1  Player1  Punch

Notice that player is split up as it is not our joining variable and needs to be recorded for both sides of the DataFrames.

The next type of **Merge** is the Left **Merge**. This **Merge** keeps all of the values from the DataFrame on the left and includes only the relevant variables from the dataframe on the right where they match up.

In [49]:
df1.merge(df2, on='Title',how='left')

Player_x  Points  Title Player_y  Power
0  Player1       8  Game1  Player1  Punch
1  Player2       9  Game2      NaN    NaN
2  Player3       6  Game3      NaN    NaN

Notice the NaN for the values missing in DataFrame 2 that could not be merged into DataFrame 1.

The next type of **Merge** is the Right **Merge**. It is very similar to the left join with the main change being that the data from the DataFrame on the right is kept but the data from the left dataframe will be joined where applicable.

In [50]:
df1.merge(df2, on='Title',how='right')

Player_x  Points  Title Player_y  Power
0  Player1     8.0  Game1  Player1  Punch
1      NaN     NaN  Game5  Player5   Kick
2      NaN     NaN  Game6  Player6  Elbow

The final **Merge** is the Outer **Merge**. With this **Merge** Pandas will **Merge** both DataFrames together on the joining variables and the non applicable variables will contain a NaN value.

In [51]:
df1.merge(df2, on='Title',how='outer')

Player_x  Points  Title Player_y  Power
0  Player1     8.0  Game1  Player1  Punch
1  Player2     9.0  Game2      NaN    NaN
2  Player3     6.0  Game3      NaN    NaN
3      NaN     NaN  Game5  Player5   Kick
4      NaN     NaN  Game6  Player6  Elbow

## Join Function

When we used the Merge function it was required for the data to have one similar attribute to join the two datasets however the join function merges the data on the index of the data and does not require a common variable to perform the merge. 

Let's start with creating two DataFrames that will be joined together at a later stage.

In [52]:
#Data frame 3 creation
player = ['Player1','Player2','Player3']
point=[8,9,6]
title=['Game1','Game2','Game3']

df3 = pd.DataFrame({'Player':player,'Points':point,'Title':title},index=['L1','L2','L3'])
df3

Player  Points  Title
L1  Player1       8  Game1
L2  Player2       9  Game2
L3  Player3       6  Game3

In [53]:
#Dataframe 4 creation

player = ['Player1','Player5','Player6']
power = ['Punch','Kick','Elbow']
title=['Game1','Game5','Game6']
df4 = pd.DataFrame({'Players':player,'Power':power,'Titles':title},index=['L2','L3','L4'])
df4

Players  Power Titles
L2  Player1  Punch  Game1
L3  Player5   Kick  Game5
L4  Player6  Elbow  Game6

The first **Join** we will explore is the inner **Join**. Very much like the **Merge** function this will only combine the common indexes. 

One of the key differences between **Join** and **Merge** functions is the column names must not be the same in either DataFrame for the **Join** function. So in the example you will see the title and player names are different between the two DataFrames.

In [54]:
df3.join(df4,how="inner")

Player  Points  Title  Players  Power Titles
L2  Player2       9  Game2  Player1  Punch  Game1
L3  Player3       6  Game3  Player5   Kick  Game5

The next **Join** is the left **Join**. Again like the **Merge** function of the same name it will keep all of the values in the left DataFrame and only join the matching data from the right DataFrame.

In [55]:
df3.join(df4,how="left")

Player  Points  Title  Players  Power Titles
L1  Player1       8  Game1      NaN    NaN    NaN
L2  Player2       9  Game2  Player1  Punch  Game1
L3  Player3       6  Game3  Player5   Kick  Game5

The final **Join** is the outer **Join**. Just like before this **Join** will include all of the data from both datasets with NaN values filling in the spaces that is not common when we join the index.

In [56]:
df3.join(df4,how="outer")

Player  Points  Title  Players  Power Titles
L1  Player1     8.0  Game1      NaN    NaN    NaN
L2  Player2     9.0  Game2  Player1  Punch  Game1
L3  Player3     6.0  Game3  Player5   Kick  Game5
L4      NaN     NaN    NaN  Player6  Elbow  Game6

## Concaternations

**Concaternations** is the final method to combine two DataSets together. This does not perform a **Join** action but rather puts the two datasets on top of each other.

Using the same two DataSets as before we will **Concatenate** them and see the final results. Please note the inclusion of the NaN values.

In [57]:
pd.concat([df3,df4],sort=True)

Player  Players  Points  Power  Title Titles
L1  Player1      NaN     8.0    NaN  Game1    NaN
L2  Player2      NaN     9.0    NaN  Game2    NaN
L3  Player3      NaN     6.0    NaN  Game3    NaN
L2      NaN  Player1     NaN  Punch    NaN  Game1
L3      NaN  Player5     NaN   Kick    NaN  Game5
L4      NaN  Player6     NaN  Elbow    NaN  Game6

## Importing and analysing a dataset

One of the key elements that we need to look at with any data science project is the importation of data we will cover this in the next section.

The first import function will be how to import a csv file using the pandas function:

In [58]:
data_in =pd.read_csv('Ratings.csv')

data_in

Const  Your Rating  ... Release Date                     Directors
0    tt0100157            8  ...   1990-11-29                    Rob Reiner
1    tt0100405            7  ...   1990-03-23                Garry Marshall
2    tt1007029            7  ...   2011-12-26                Phyllida Lloyd
3    tt1010048            7  ...   2008-08-30  Loveleen Tandan, Danny Boyle
4    tt0101272            7  ...   1991-11-19              Barry Sonnenfeld
..         ...          ...  ...          ...                           ...
701  tt0099088            8  ...   1990-05-25               Robert Zemeckis
702  tt0993846            9  ...   2013-12-09               Martin Scorsese
703  tt0099487            8  ...   1990-12-06                    Tim Burton
704  tt0099653            8  ...   1990-07-13                  Jerry Zucker
705  tt0099864            7  ...   1990-11-18                           NaN

[706 rows x 13 columns]

With the newly imported data we can find out how python has imported it by checking the **type**.

In [59]:
type(data_in)

pandas.core.frame.DataFrame

There may be moments when we need to look at certain parts of the data but don't want to view the full data set. 

There is where the head and tail comes in use. You can either keep the default of looking at the 5 values or parse a number for the number of rows you want to look at.

If it is the start of the data you will use the head function and the bottom of the data you will use tail.

In [60]:
data_in.head()

Const  Your Rating  ... Release Date                     Directors
0  tt0100157            8  ...   1990-11-29                    Rob Reiner
1  tt0100405            7  ...   1990-03-23                Garry Marshall
2  tt1007029            7  ...   2011-12-26                Phyllida Lloyd
3  tt1010048            7  ...   2008-08-30  Loveleen Tandan, Danny Boyle
4  tt0101272            7  ...   1991-11-19              Barry Sonnenfeld

[5 rows x 13 columns]

The default **head** function will show up a default of the top 5 values however we can parse in a value to increase the display (For instance 10 below):

In [61]:
data_in.head(10)

Const  Your Rating  ... Release Date                     Directors
0  tt0100157            8  ...   1990-11-29                    Rob Reiner
1  tt0100405            7  ...   1990-03-23                Garry Marshall
2  tt1007029            7  ...   2011-12-26                Phyllida Lloyd
3  tt1010048            7  ...   2008-08-30  Loveleen Tandan, Danny Boyle
4  tt0101272            7  ...   1991-11-19              Barry Sonnenfeld
5  tt1013752            7  ...   2009-03-12                    Justin Lin
6  tt1013977            9  ...   2007-09-30                  Tony Goldwyn
7  tt1013981            9  ...   2007-10-07                  Marcos Siega
8  tt1013982            8  ...   2007-10-14                  Tony Goldwyn
9  tt1013983            8  ...   2007-10-21                    Nick Gomez

[10 rows x 13 columns]

The same applys to the **tail** function:

In [62]:
data_in.tail()

Const  Your Rating  ... Release Date        Directors
701  tt0099088            8  ...   1990-05-25  Robert Zemeckis
702  tt0993846            9  ...   2013-12-09  Martin Scorsese
703  tt0099487            8  ...   1990-12-06       Tim Burton
704  tt0099653            8  ...   1990-07-13     Jerry Zucker
705  tt0099864            7  ...   1990-11-18              NaN

[5 rows x 13 columns]

In [63]:
data_in.tail(10)

Const  Your Rating  ... Release Date        Directors
696  tt0098084            7  ...   1989-04-21     Mary Lambert
697  tt0985699            9  ...   2008-12-25     Bryan Singer
698  tt0986263            8  ...   2009-09-24  Jonathan Mostow
699  tt0988045            8  ...   2009-12-14      Guy Ritchie
700  tt0990407            5  ...   2011-01-12    Michel Gondry
701  tt0099088            8  ...   1990-05-25  Robert Zemeckis
702  tt0993846            9  ...   2013-12-09  Martin Scorsese
703  tt0099487            8  ...   1990-12-06       Tim Burton
704  tt0099653            8  ...   1990-07-13     Jerry Zucker
705  tt0099864            7  ...   1990-11-18              NaN

[10 rows x 13 columns]

If we need more details about the number of rows and columns that have been imported in the data we can use the **shape** function. This will return a set of values to indicate the number of rows and columns present.

In [64]:
data_in.shape

(706, 13)

This means we have 706 rows and 13 columns in the data Set.

But this may not be enough information for us. We may want to check the number of values per variable as well as the type that each variable has been imported as. This is where the **info** function is highly useful.

In [65]:
data_in.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 706 entries, 0 to 705
Data columns (total 13 columns):
Const             706 non-null object
Your Rating       706 non-null int64
Date Rated        706 non-null object
Title             706 non-null object
URL               706 non-null object
Title Type        706 non-null object
IMDb Rating       706 non-null float64
Runtime (mins)    706 non-null int64
Year              706 non-null int64
Genres            706 non-null object
Num Votes         706 non-null int64
Release Date      706 non-null object
Directors         685 non-null object
dtypes: float64(1), int64(4), object(8)
memory usage: 71.8+ KB


At the dtypes info we can see we have 1 floating value, 4 interger columns and 8 objects in this dataset - this is a view of the data types in the current dataset. 

The next bit will cover the various basic Statistic measures that we may want to look at. The first of which will be two locational measures (The **mean** and the **median**)

In [66]:
data_in.mean()

Your Rating            7.174221
IMDb Rating            7.043909
Runtime (mins)       111.237960
Year                2006.871105
Num Votes         322995.311615
dtype: float64

In [67]:
data_in.median()

Your Rating            7.0
IMDb Rating            7.0
Runtime (mins)       110.0
Year                2009.0
Num Votes         231925.0
dtype: float64

Now that we have a few measures of location, we should start looking at some of the measures of spread to see how we can analyse the data. One of the first methods is the standard Deviation as well as the Maximum and Minimum Values.

In [68]:
data_in.std()

Your Rating            1.178754
IMDb Rating            0.953678
Runtime (mins)        29.839857
Year                   9.197692
Num Votes         309869.726427
dtype: float64

In [69]:
data_in.max()

Const                                         tt9358204
Your Rating                                          10
Date Rated                                   2019-12-20
Title                                          Æon Flux
URL               https://www.imdb.com/title/tt9358204/
Title Type                                        video
IMDb Rating                                         9.7
Runtime (mins)                                      361
Year                                               2019
Genres                                 Sci-Fi, Thriller
Num Votes                                       2188152
Release Date                                 2019-12-16
dtype: object

In [70]:
data_in.min()

Const                                                     tt0055928
Your Rating                                                       2
Date Rated                                               2014-04-19
Title                                          (500) Days of Summer
URL                           https://www.imdb.com/title/tt0055928/
Title Type                                                    movie
IMDb Rating                                                     3.1
Runtime (mins)                                                    5
Year                                                           1962
Genres            Action, Adventure, Biography, Drama, History, ...
Num Votes                                                        10
Release Date                                             1962-10-05
dtype: object

Sometimes the measures of location and spread are not enough to indicate the underlying issues with the data - for instance missing data in the variables. This is where count can assist:

In [71]:
#count non null:

data_in.count()

Const             706
Your Rating       706
Date Rated        706
Title             706
URL               706
Title Type        706
IMDb Rating       706
Runtime (mins)    706
Year              706
Genres            706
Num Votes         706
Release Date      706
Directors         685
dtype: int64

And finally we may want to have a view of the full dataset. This is where the **describe** function comes in - it will offer various statistics for all of the quantitative variables in the dataset.

In [72]:
data_in.describe()

Your Rating  IMDb Rating  Runtime (mins)         Year     Num Votes
count   706.000000   706.000000      706.000000   706.000000  7.060000e+02
mean      7.174221     7.043909      111.237960  2006.871105  3.229953e+05
std       1.178754     0.953678       29.839857     9.197692  3.098697e+05
min       2.000000     3.100000        5.000000  1962.000000  1.000000e+01
25%       7.000000     6.400000       97.000000  2003.000000  1.197800e+05
50%       7.000000     7.000000      110.000000  2009.000000  2.319250e+05
75%       8.000000     7.700000      125.000000  2014.000000  4.264912e+05
max      10.000000     9.700000      361.000000  2019.000000  2.188152e+06

## Cleaning the Data set

Using the dataset as it comes may not be the ideal way to analyse without doing a bit of data cleaning. We should first start with a cleaning of the dataset.

First part is rename the columns.

In [73]:
data_in = data_in.rename(columns={'URL':'Web Address'})
data_in.head()

Const  Your Rating  ... Release Date                     Directors
0  tt0100157            8  ...   1990-11-29                    Rob Reiner
1  tt0100405            7  ...   1990-03-23                Garry Marshall
2  tt1007029            7  ...   2011-12-26                Phyllida Lloyd
3  tt1010048            7  ...   2008-08-30  Loveleen Tandan, Danny Boyle
4  tt0101272            7  ...   1991-11-19              Barry Sonnenfeld

[5 rows x 13 columns]

With Bigger data sets we may end up in a situation where we have variables that won't be usable in any future analysis we have planned. This is where we can use the drop columns to reduce the size of the full dataset.

In [74]:
data_in=data_in.drop(columns=['Const'])
data_in.head()

Your Rating  Date Rated  ... Release Date                     Directors
0            8  2014-04-21  ...   1990-11-29                    Rob Reiner
1            7  2019-06-16  ...   1990-03-23                Garry Marshall
2            7  2014-04-22  ...   2011-12-26                Phyllida Lloyd
3            7  2015-03-13  ...   2008-08-30  Loveleen Tandan, Danny Boyle
4            7  2014-04-21  ...   1991-11-19              Barry Sonnenfeld

[5 rows x 12 columns]

Once we have a working dataset and before we need to decide on the model to be applying to the data. we will need to determine the relationships. One of the most basic ways of doing this is through the Correlation matrix:

In [75]:
data_in[['Your Rating','IMDb Rating','Runtime (mins)','Year']].corr()


Your Rating  IMDb Rating  Runtime (mins)      Year
Your Rating        1.000000     0.586351       -0.009205  0.066704
IMDb Rating        0.586351     1.000000       -0.014269 -0.155935
Runtime (mins)    -0.009205    -0.014269        1.000000  0.016268
Year               0.066704    -0.155935        0.016268  1.000000

In Some instances the data we receive may be imported with the wrong variable types and need to be changed before we can include them in analysis. First lets get the type of all the variables and then change year to type Float.

In [76]:
#changing the data type:
#of the year to float

data_in.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 706 entries, 0 to 705
Data columns (total 12 columns):
Your Rating       706 non-null int64
Date Rated        706 non-null object
Title             706 non-null object
Web Address       706 non-null object
Title Type        706 non-null object
IMDb Rating       706 non-null float64
Runtime (mins)    706 non-null int64
Year              706 non-null int64
Genres            706 non-null object
Num Votes         706 non-null int64
Release Date      706 non-null object
Directors         685 non-null object
dtypes: float64(1), int64(4), object(7)
memory usage: 66.3+ KB


In [77]:
data_in.Year = data_in.Year.astype(float)
data_in.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 706 entries, 0 to 705
Data columns (total 12 columns):
Your Rating       706 non-null int64
Date Rated        706 non-null object
Title             706 non-null object
Web Address       706 non-null object
Title Type        706 non-null object
IMDb Rating       706 non-null float64
Runtime (mins)    706 non-null int64
Year              706 non-null float64
Genres            706 non-null object
Num Votes         706 non-null int64
Release Date      706 non-null object
Directors         685 non-null object
dtypes: float64(2), int64(3), object(7)
memory usage: 66.3+ KB


## Manipulating the data

Now that we have covered most of the ways we have to combine and fix issues in the data it is now time to look at the ways we can subset the data and either select rows or columns.

The first of which is to select specific columns using the **iloc** function - this function only passes in numerical values in the selection. For the example we want to select the third column so we will put a 2 in the column section of the selection criteria [,].

Since we are wanting the 3 column we will use the **iloc[,]** to obtain this information, it is important to note that the inputs for the **[,]** is row,column. We are selecting all rows so will use ":" in the row request and 2 in the columns.

In [78]:

data_in.iloc[:,2]


0                           Misery
1                     Pretty Woman
2                    The Iron Lady
3              Slumdog Millionaire
4                The Addams Family
                  ...             
701    Back to the Future Part III
702        The Wolf of Wall Street
703            Edward Scissorhands
704                          Ghost
705                             It
Name: Title, Length: 706, dtype: object

Now that we have explored a specific way to select only on column the next consideration is how to select sections of rows as well as sections of columns. For this section we will first select the first 5 rows from colums 5. This requires us to use a " : " to indicate to select a range.

In [79]:
data_in.iloc[0:5,4]

0    movie
1    movie
2    movie
3    movie
4    movie
Name: Title Type, dtype: object

Using what we have learnt lets now select all of the rows and columns from the dataset.

In [80]:
data_in.iloc[:,:]

Your Rating  Date Rated  ... Release Date                     Directors
0              8  2014-04-21  ...   1990-11-29                    Rob Reiner
1              7  2019-06-16  ...   1990-03-23                Garry Marshall
2              7  2014-04-22  ...   2011-12-26                Phyllida Lloyd
3              7  2015-03-13  ...   2008-08-30  Loveleen Tandan, Danny Boyle
4              7  2014-04-21  ...   1991-11-19              Barry Sonnenfeld
..           ...         ...  ...          ...                           ...
701            8  2015-02-27  ...   1990-05-25               Robert Zemeckis
702            9  2014-06-02  ...   2013-12-09               Martin Scorsese
703            8  2015-02-20  ...   1990-12-06                    Tim Burton
704            8  2015-02-20  ...   1990-07-13                  Jerry Zucker
705            7  2015-02-20  ...   1990-11-18                           NaN

[706 rows x 12 columns]

Let's now use the selection to create a subset that will reduce the elements in the dataset based on an exclusion of certain variables rather than selecting only certain ones. Since we have a starting location we won't include an ending point in the request.

In [81]:
data_in.iloc[5:,5:]

IMDb Rating  Runtime (mins)  ...  Release Date        Directors
5            6.6             107  ...    2009-03-12       Justin Lin
6            8.8              53  ...    2007-09-30     Tony Goldwyn
7            8.6              56  ...    2007-10-07     Marcos Siega
8            8.7              52  ...    2007-10-14     Tony Goldwyn
9            8.5              53  ...    2007-10-21       Nick Gomez
..           ...             ...  ...           ...              ...
701          7.4             118  ...    1990-05-25  Robert Zemeckis
702          8.2             180  ...    2013-12-09  Martin Scorsese
703          7.9             105  ...    1990-12-06       Tim Burton
704          7.0             127  ...    1990-07-13     Jerry Zucker
705          6.8             192  ...    1990-11-18              NaN

[701 rows x 7 columns]

If we want to select the columns by name rather than number we can use the location command or the loc.

This also takes the input of [row,column] with the big difference of the columns passed through as strings.

Again we will go through a similar exercise as the iloc with the first selection of a full column, then a section of the column

In [82]:
data_in.loc[:,'Year']

0      1990.0
1      1990.0
2      2011.0
3      2008.0
4      1991.0
        ...  
701    1990.0
702    2013.0
703    1990.0
704    1990.0
705    1990.0
Name: Year, Length: 706, dtype: float64

In [83]:
data_in.loc[:6,'Year']

0    1990.0
1    1990.0
2    2011.0
3    2008.0
4    1991.0
5    2009.0
6    2007.0
Name: Year, dtype: float64

## Other Data Maipulation

The next section will cover a few other elements of data manipulation that may become useful at a later stage.

The first of which is inserting values into a data frame, for this example we can insert a constant value of 1 in a row.

In [84]:
data_in['const']=1
data_in.head()

Your Rating  Date Rated  ...                     Directors const
0            8  2014-04-21  ...                    Rob Reiner     1
1            7  2019-06-16  ...                Garry Marshall     1
2            7  2014-04-22  ...                Phyllida Lloyd     1
3            7  2015-03-13  ...  Loveleen Tandan, Danny Boyle     1
4            7  2014-04-21  ...              Barry Sonnenfeld     1

[5 rows x 13 columns]

The next interesting manipulation is to apply a function to the full dataset row by row. In the example below we are going to create a new function that will multiple the values by and then use the newly created function to apply it to all of the rows in the dataset.

In [85]:
f = lambda x:x*2

data_in['const']=data_in['const'].apply(f)
data_in.head()

Your Rating  Date Rated  ...                     Directors const
0            8  2014-04-21  ...                    Rob Reiner     2
1            7  2019-06-16  ...                Garry Marshall     2
2            7  2014-04-22  ...                Phyllida Lloyd     2
3            7  2015-03-13  ...  Loveleen Tandan, Danny Boyle     2
4            7  2014-04-21  ...              Barry Sonnenfeld     2

[5 rows x 13 columns]

In the example above we create a **lamdba** function, this is a temporary annonymous function that will take one input (here it is x) and do a quick function on it.

The next application that is avaliable to us is a way to sort the data by certain columns.

In [86]:
data_in.sort_values(by='Year')

Your Rating  Date Rated  ...                     Directors const
601            7  2015-03-13  ...                 Terence Young     2
604            8  2015-03-13  ...                  Guy Hamilton     2
607            7  2015-02-27  ...                 Terence Young     2
610            8  2015-02-20  ...                           NaN     2
612            9  2015-05-31  ...          Francis Ford Coppola     2
..           ...         ...  ...                           ...   ...
652            7  2019-07-14  ...  Dennis Widmyer, Kevin Kölsch     2
24             8  2019-10-07  ...                 Rebecca Sugar     2
264            8  2019-06-28  ...               Dexter Fletcher     2
602            9  2019-11-10  ...                 Mike Flanagan     2
609            7  2019-09-30  ...                Chad Stahelski     2

[706 rows x 13 columns]

In [87]:
data_in.sort_values(by='Year',ascending=False)

Your Rating  Date Rated  ...             Directors const
288            5  2019-11-09  ...          F. Gary Gray     2
606            7  2019-05-27  ...         Rob Letterman     2
442            7  2019-05-02  ...         Mike Mitchell     2
324            8  2019-12-20  ...           J.J. Abrams     2
24             8  2019-10-07  ...         Rebecca Sugar     2
..           ...         ...  ...                   ...   ...
612            9  2015-05-31  ...  Francis Ford Coppola     2
610            8  2015-02-20  ...                   NaN     2
607            7  2015-02-27  ...         Terence Young     2
604            8  2015-03-13  ...          Guy Hamilton     2
601            7  2015-03-13  ...         Terence Young     2

[706 rows x 13 columns]

As much as it is useful for us to subset the data it is usually more useful for us to filter the data based on some criteria. Look at the examples below on how we create filters and apply them.

In [88]:
data_in['Year']<1999

0       True
1       True
2      False
3      False
4       True
       ...  
701     True
702    False
703     True
704     True
705     True
Name: Year, Length: 706, dtype: bool

In the example above we have shown the results of the evaluation but have not applied the filter to the dataset. Let's now apply the filter to the associated dataset.

In [89]:
filter1 = data_in['Year']<1999
filtered_new = data_in[filter1]
filtered_new.head()

Your Rating  Date Rated  ...                  Directors const
0             8  2014-04-21  ...                 Rob Reiner     2
1             7  2019-06-16  ...             Garry Marshall     2
4             7  2014-04-21  ...           Barry Sonnenfeld     2
12            7  2015-05-26  ...  Gary Trousdale, Kirk Wise     2
15            8  2015-09-30  ...           Steven Spielberg     2

[5 rows x 13 columns]

In the previous example we only filtered on one condition, let's now create a filter based on multiple variables:

In [90]:
filter2 = (data_in['Year']<1999) & (data_in['Your Rating']>7)
filter_review=data_in[filter2]
filter_review.head()

Your Rating  Date Rated  ...         Directors const
0             8  2014-04-21  ...        Rob Reiner     2
15            8  2015-09-30  ...  Steven Spielberg     2
17            8  2014-04-20  ...    Jonathan Demme     2
18            8  2014-04-21  ...     James Cameron     2
22            8  2015-05-26  ...         Nick Park     2

[5 rows x 13 columns]